Source/Scene/ModelInstanceCollection.js

```diff
@@ -0,0 +1,844 @@
+                                    );
+                                    //>>includeEnd('debug');
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        collection._instancedUniformsByProgram = instancedUniformsByProgram;
+        return instancedUniformsByProgram[programName];
+    }
+
+    var vertexShaderCached;
+
+    function getVertexShaderCallback(collection) {
+        return function(vs, programName) {
+            var instancedUniforms = getInstancedUniforms(collection, programName);
+            var dynamic = collection._dynamic;
+            var regex;
+
+            var renamedSource = ShaderSource.replaceMain(vs, 'czm_old_main');
+
+            // TODO : can I add all these global vars and just let the shader compiler optimize them out
+            var globalVarsHeader = '';
+            var globalVarsMain = '';
+            for (var uniform in instancedUniforms) {
+                if (instancedUniforms.hasOwnProperty(uniform)) {
+                    var semantic = instancedUniforms[uniform];
+                    var varName;
+                    if (semantic === 'MODELVIEW') {
+                        varName = 'czm_instanced_modelView';
+                    } else if (semantic === 'MODELVIEWPROJECTION') {
+                        varName = 'czm_instanced_modelViewProjection';
+                        globalVarsHeader += 'mat4 czm_instanced_modelViewProjection;\n';
+                        globalVarsMain += 'czm_instanced_modelViewProjection = czm_projection * czm_instanced_modelView;\n';
+                    } else if (semantic === 'MODELVIEWINVERSETRANSPOSE') {
+                        varName = 'czm_instanced_modelViewInverseTranspose';
+                        globalVarsHeader += 'mat3 czm_instanced_modelViewInverseTranspose;\n';
+                        globalVarsMain += 'czm_instanced_modelViewInverseTranspose = mat3(czm_instanced_modelView);\n';
+                    }
+
+                    // Remove the uniform declaration
+                    regex = new RegExp('uniform.*' + uniform + '.*');
+                    renamedSource = renamedSource.replace(regex, '');
+
+                    // Replace all occurrences of the uniform with the global variable
+                    regex = new RegExp(uniform + '\\b', 'g');
+                    renamedSource = renamedSource.replace(regex, varName);
+                }
+            }
+
+            var dynamicDefine = dynamic ? '#define DYNAMIC\n' : '';
+
+            var newMain =
+                dynamicDefine +
+                globalVarsHeader +
+                'attribute vec4 czm_modelMatrixRow0;\n' +
+                'attribute vec4 czm_modelMatrixRow1;\n' +
+                'attribute vec4 czm_modelMatrixRow2;\n' +
+                'attribute float a_batchId;\n' + // TODO : what if the gltf has this attribute already?
```

Comment:
I didn't because the batched model tile expects the gltf model to have an attribute called a_batchId. Both batched tiles and instances tiles use `Cesium3DTileBatchData.getVertexShaderCallback`, which expects`a_batchId`.